Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -108,3 +133,116 @@ extension KeyLogCallbackManager {
         self.callback(self.scratchBuffer)
     }
 }
+
+
+/// A struct that provides helpers for working with a NIOSSLCustomVerificationCallback.
+internal struct CustomVerifyManager {
+    private var callback: CallbackType
+
+    private var result: PendingResult = .notStarted
+
+    init(callback: @escaping NIOSSLCustomVerificationCallback) {
+        self.callback = .public(callback)
+    }
+
+    init(callback: @escaping InternalCallback) {
+        self.callback = .internal(callback)
+    }
+}
+
+
+extension CustomVerifyManager {
+    private enum PendingResult {
+        case notStarted
+
+        case pendingResult
+
+        case complete(NIOSSLVerificationResult)
+    }
+}
+
+
+extension CustomVerifyManager {
+    mutating func process(on connection: SSLConnection) -> ssl_verify_result_t {
+        // First, check if we have a result.
+        switch self.result {
+        case .complete(.certificateVerified):
+            return ssl_verify_ok
+        case .complete(.failed):
+            return ssl_verify_invalid
+        case .pendingResult:
+            // Ask me again.
+            return ssl_verify_retry
+        case .notStarted:
+            // The rest of this method handles this case.
+            break
+        }
+
+        self.result = .pendingResult
+
+        // Ok, no result. We need a promise for the user to use to supply a result.
+        guard let eventLoop = connection.eventLoop else {
+            // No event loop. We cannot possibly be negotiating here.
+            preconditionFailure("No event loop present")
+        }
+
+        let promise = eventLoop.makePromise(of: NIOSSLVerificationResult.self)
+
+        // We need to attach our "do the thing" callback. This will always invoke the "ask me again" API, and it will do so in a separate
+        // event loop tick to avoid awkward re-entrancy with this method.
+        promise.futureResult.whenComplete { result in
+            // When we complete here we need to set our result state, and then ask to respin certificate verification.
+            // If we can't respin verification because we've dropped the parent handler, that's fine, no harm no foul.
+            // For that reason, we tolerate both the verify manager and the parent handler being nil.
+            eventLoop.execute {
+                // Note that we don't close over self here: that's to deal with the fact that this is a struct, and we don't want to
+                // escape the mutable ownership of self.
+                connection.customVerificationManager?.result = .complete(NIOSSLVerificationResult(result))
```

Comment:
could you assert here that we're actually still in the state that we want to be in before overwriting the values here?